File: lib/sync.ts
// Incremental sync orchestrator. Reuses the existing building blocks:
//   1. load prior metadata_hash map from the DB
//   2. dumpTypes(incremental) — rewrite only new/changed articles, leave unchanged
//   3. enrichDump — bodies only for the rewritten (resumable skips already-bodied)
//   4. trackDump — update the DB + its changes/runs tables
//   5. DETECT + REPORT upstream deletions (DB ids absent from this run's Coveo set);
//      never removes them (use `reconcile --apply` for that)
//   6. flush the changelog (added/edited/body-* this run; deletions recorded as
//      "detected", source=sync)
//
// Deletion detection requires the full current id set, so it only runs under
// --all (allTime). A --days / --since-last-run window classifies adds/edits but
// cannot see deletions.

import { type Logger, NULL_LOGGER } from "./logger.ts";
import { CoveoClient } from "./coveo/client.ts";
import { HttpClient } from "./http/fetcher.ts";
import type { TypeConfig } from "./config/types.ts";
import { dumpTypes } from "./dump.ts";
import { enrichDump } from "./enrich/driver.ts";
import { loadHashIndex, loadIdsByType, trackDump } from "./track/db.ts";
import { Changelog } from "./changelog.ts";
import { mergePending } from "./staging.ts";

// type keys whose body lives off-API and so benefits from enrich.
const ENRICHABLE = new Set([
  "Bug_Tracker",
  "Manual",
  "Release_Note",
  "Supplemental_Document",
  "F5_GitHub",
]);

export interface SyncOpts {
  client: CoveoClient;
  http: HttpClient;
  typeConfigs: Record<string, Partial<TypeConfig>>;
  typeKeys: string[];
  descriptions: Record<string, string>;
  outDir: string;
  db?: string;
  mode: string; // "all" | "days=N" | "since-last-run"
  allTime: boolean; // true => deletion detection is valid
  cutoffMs: number;
  endMs: number;
  nowMs: number;
  pageSize: number;
  limit: number;
  configPath: string;
  enrich: boolean;
  githubToken?: string;
  concurrency: number;
  delayMs: number;
  changelogPath: string | null;
  dryRun: boolean;
  /** approval gate ON (default): edited articles are staged to _pending/ for review
   *  instead of overwriting live data. */
  approval: boolean;
  /** bypass (--yes): overwrite edited articles in place, archiving the replaced file
   *  to _replaced/ first. */
  archiveOnOverwrite?: boolean;
  logger?: Logger;
}

export interface SyncResult {
  runId: string;
  mode: string;
  dryRun: boolean;
  written: number;
  skipped: number;
  added: number;
  edited: number;
  bodyAdded: number;
  bodyChanged: number;
  bodyError: number;
  /** edited articles staged to _pending/ this run (awaiting `f5kb approve`). */
  staged: number;
  deletionsDetected: number;
  deletions: Record<string, string[]>; // typeKey -> ids detected as deleted upstream
  deletionDetectionRan: boolean;
  changelogPath: string | null;
}

export async function syncDump(opts: SyncOpts): Promise<SyncResult> {
  const log = opts.logger ?? NULL_LOGGER;
  const runId = new Date(opts.nowMs).toISOString();
  const dbPath = opts.db ?? `${opts.outDir.replace(/\/+$/, "")}/../articles.db`;
  // In dry-run, the changelog still tallies counts but writes no file (path=null).
  const changelog = new Changelog(opts.dryRun ? null : opts.changelogPath, runId);

  const priorHashes = await loadHashIndex(dbPath);
  log.info(`Loaded ${priorHashes.size} prior hash(es) from ${dbPath}`);

  // 1+2. incremental dump (only new/changed written).
  const dump = await dumpTypes(opts.client, {
    typeConfigs: opts.typeConfigs,
    typeKeys: opts.typeKeys,
    descriptions: opts.descriptions,
    outDir: opts.outDir,
    allTime: opts.allTime,
    mode: opts.mode,
    cutoffMs: opts.cutoffMs,
    endMs: opts.endMs,
    nowMs: opts.nowMs,
    pageSize: opts.pageSize,
    limit: opts.limit,
    configPath: opts.configPath,
    logger: log,
    incremental: true,
    priorHashes,
    changelog,
    dryRun: opts.dryRun,
    approval: opts.approval,
    archiveOnOverwrite: opts.archiveOnOverwrite,
  });

  // 3. enrich only the rewritten (resumability skips already-bodied) — live + writes,
  //    so skipped in dry-run.
  if (opts.enrich && !opts.dryRun) {
    const enrichTypes = opts.typeKeys.filter((t) => ENRICHABLE.has(t));
    if (enrichTypes.length) {
      await enrichDump({
        dump: opts.outDir,
        types: enrichTypes,
        http: opts.http,
        githubToken: opts.githubToken,
        concurrency: opts.concurrency,
        delayMs: opts.delayMs,
        limit: null,
        refetch: false,
        refetchErrors: false,
        logger: log.child("enrich"),
        changelog,
      });
    }
  }

  // 3b. record the staged overwrites in the pending manifest (after enrich filled
  //     their bodies). Only meaningful when the gate is on and not a dry run.
  if (opts.approval && !opts.dryRun && dump.pending.length) {
    await mergePending(opts.outDir, dump.pending, runId);
    log.warn(
      `${dump.pending.length} edited article(s) STAGED for review (not applied). ` +
        `Inspect ${opts.outDir}/_pending/ then run: f5kb approve`,
    );
  }

  // 4. track (update DB) — skipped in dry-run. Indexes the LIVE dump only; staged
  //    (_pending) edits are excluded until approved.
  if (!opts.dryRun) {
    await trackDump({ dump: opts.outDir, db: dbPath, types: opts.typeKeys, runId, logger: log });
  }

  // 5. deletion DETECTION (only valid with the full current id set, i.e. --all).
  const deletions: Record<string, string[]> = {};
  let deletionsDetected = 0;
  const deletionDetectionRan = opts.allTime;
  if (deletionDetectionRan) {
    const docTypes = new Map<string, string>();
    for (const k of opts.typeKeys) {
      const dt = opts.typeConfigs[k]?.documentType;
      if (dt) docTypes.set(k, dt);
    }
    const dbIdsByType = await loadIdsByType(dbPath, [...docTypes.values()]);
    for (const [typeKey, documentType] of docTypes) {
      const cur = dump.currentIds.get(typeKey) ?? new Set<string>();
      const gone = (dbIdsByType.get(documentType) ?? []).filter((id) => !cur.has(id));
      if (gone.length) {
        deletions[typeKey] = gone;
        deletionsDetected += gone.length;
        for (const id of gone) {
          changelog.record({
            op: "deleted",
            documentType,
            id,
            source: "sync",
            detail: "detected upstream (not removed; run `reconcile --apply` to remove)",
          });
        }
      }
    }
    if (deletionsDetected) {
      log.warn(
        `${deletionsDetected} upstream deletion(s) detected (reported, NOT removed). ` +
          `Run: f5kb reconcile --apply  to remove them.`,
      );
    }
  }

  await changelog.flush();

  const by = changelog.byOp();
  return {
    runId,
    mode: opts.mode,
    dryRun: opts.dryRun,
    written: dump.manifest.reduce((a, m) => a + m.written, 0),
    skipped: dump.manifest.reduce((a, m) => a + m.skipped, 0),
    added: by["added"] ?? 0,
    edited: by["edited"] ?? 0,
    bodyAdded: by["body-added"] ?? 0,
    bodyChanged: by["body-changed"] ?? 0,
    bodyError: by["body-error"] ?? 0,
    staged: dump.manifest.reduce((a, m) => a + m.staged, 0),
    deletionsDetected,
    deletions,
    deletionDetectionRan,
    changelogPath: opts.dryRun ? null : opts.changelogPath,
  };
}

File: f5kb/cmd/sync.py
"""f5kb sync subcommand."""

from __future__ import annotations

import os
import sys
import time

import click

from f5kb.config.loader import load_config
from f5kb.coveo.aura import fetch_coveo_config, refresh_config
from f5kb.coveo.client import CoveoClient
from f5kb.http.fetcher import HttpClient
from f5kb.lib.changelog import changelog_path_from_flag
from f5kb.lib.sync import sync_dump
from f5kb.track.db import load_last_run_at


@click.command()
@click.option("--all", "all_time", is_flag=True)
@click.option("--days", "days_str", default=None)
@click.option("--since-last-run", is_flag=True)
@click.option("--types", default=None)
@click.option("--exclude-types", default=None)
@click.option("--out", "out_dir", default="outputs/dump", show_default=True)
@click.option("--config", "config_path", default="config.yaml", show_default=True)
@click.option("--db", default=None)
@click.option("--no-enrich", is_flag=True)
@click.option("--changelog", "changelog_flag", default="", is_flag=False, flag_value="")
@click.option("--no-changelog", is_flag=True)
@click.option("--dry-run", is_flag=True)
@click.option("--yes", is_flag=True)
@click.option("--page-size", type=int, default=200, show_default=True)
@click.option("--limit", type=int, default=0)
@click.option("--concurrency", type=int, default=4, show_default=True)
@click.option("--delay-ms", type=int, default=200, show_default=True)
@click.pass_context
def sync_cmd(ctx, all_time, days_str, since_last_run, types, exclude_types, out_dir,
             config_path, db, no_enrich, changelog_flag, no_changelog, dry_run, yes,
             page_size, limit, concurrency, delay_ms) -> None:
    """Incremental update: dump+enrich+track only changed; detect deletions."""
    log = ctx.obj["logger"]
    mode_count = sum([all_time, bool(days_str), since_last_run])
    if mode_count == 0:
        log.error("provide one of --all, --days=N, or --since-last-run")
        sys.exit(1)
    if mode_count > 1:
        log.error("--all, --days=N and --since-last-run are mutually exclusive")
        sys.exit(1)

    try:
        config = load_config(config_path)
    except Exception as e:
        log.error(f"could not read/parse config {config_path}: {e}")
        sys.exit(1)
    type_configs = {k: {"documentType": v.document_type, "metadata": v.metadata, "content": v.content}
                   for k, v in config.types.items()}
    all_type_keys = list(type_configs.keys())
    include_types = [t.strip() for t in types.split(",")] if types else None
    excl_types = [t.strip() for t in exclude_types.split(",")] if exclude_types else None
    type_keys = [t for t in all_type_keys if
                 (not include_types or t in include_types) and
                 (not excl_types or t not in excl_types)]
    if not type_keys:
        log.error("no type keys selected")
        sys.exit(1)

    now_ms = int(time.time() * 1000)
    db_path = db or f"{out_dir.rstrip('/')}/../articles.db"

    if since_last_run:
        last = load_last_run_at(db_path)
        if not last:
            log.error("--since-last-run: no prior run in DB; use --all or --days=N first")
            sys.exit(1)
        cutoff_ms = last["ran_at_ms"] or (now_ms - 7 * 86400000)
        mode = "since-last-run"
    elif days_str:
        try:
            d = float(days_str)
            assert d > 0
        except Exception:
            log.error("--days must be a positive number")
            sys.exit(1)
        cutoff_ms = int(now_ms - d * 86400000)
        mode = f"days={d}"
    else:
        cutoff_ms = 946684800000
        mode = "all"

    end_ms = now_ms + 86400000
    cl_path = None if no_changelog else changelog_path_from_flag(changelog_flag, out_dir)

    log.info("Fetching Coveo configuration from F5 portal...")
    try:
        coveo_config = fetch_coveo_config()
    except Exception as e:
        log.error(f"failed to fetch Coveo config: {e}")
        sys.exit(1)
    client = CoveoClient(coveo_config, logger=log.child("coveo"), refresh=lambda c: refresh_config(c))
    http = HttpClient(logger=log.child("http"))
    github_token = os.environ.get("GITHUB_TOKEN")

    result = sync_dump(
        client=client,
        http=http,
        type_configs=type_configs,
        type_keys=type_keys,
        descriptions=dict(config.field_descriptions),
        out_dir=out_dir,
        db=db_path,
        mode=mode,
        all_time=all_time,
        cutoff_ms=cutoff_ms,
        end_ms=end_ms,
        now_ms=now_ms,
        page_size=min(page_size, 500),
        limit=limit,
        config_path=config_path,
        enrich=not no_enrich,
        github_token=github_token,
        concurrency=max(1, concurrency),
        delay_ms=max(0, delay_ms),
        changelog_path=cl_path,
        dry_run=dry_run,
        approval=not yes,
        archive_on_overwrite=yes,
        logger=log,
    )

    log.info(
        f"Sync {'(dry-run) ' if dry_run else ''}done. "
        f"written={result.written} skipped={result.skipped} "
        f"added={result.added} edited={result.edited} staged={result.staged} "
        f"deletions_detected={result.deletions_detected}"
    )
    if result.changelog_path:
        log.info(f"Changelog: {result.changelog_path}")
